packages/request/lib/SessionPool.js

```diff
@@ -0,0 +1,274 @@
+//
+// SPDX-FileCopyrightText: 2020 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+'use strict'
+
+const http2 = require('http2')
+const net = require('net')
+const { globalLogger: logger } = require('@gardener-dashboard/logger')
+const Semaphore = require('./Semaphore')
+const { TimeoutError, isAbortError } = require('./errors')
+
+const {
+  NGHTTP2_CANCEL
+} = http2.constants
+
+const kSemaphore = Symbol('semaphore')
+const kTimestamp = Symbol('timestamp')
+const kTimeoutId = Symbol('timeoutId')
+const kIntervalId = Symbol('intervalId')
+
+const delay = milliseconds => new Promise(resolve => setTimeout(resolve, milliseconds))
+
+class SessionPool {
+  constructor (id) {
+    this.id = id
+    this.sessions = new Set()
+    this.tlsSession = undefined
+    this.queue = []
+  }
+
+  get options () {
+    return this.id.getOptions()
+  }
+
+  get pingInterval () {
+    return this.options.pingInterval
+  }
+
+  get keepAliveTimeout () {
+    return this.options.keepAliveTimeout
+  }
+
+  get connectTimeout () {
+    return this.options.connectTimeout
+  }
+
+  get value () {
+    let value = 0
+    for (const { [kSemaphore]: semaphore } of this.sessions) {
+      value += semaphore.value
+    }
+    return value
+  }
+
+  destroy () {
+    this.tlsSession = undefined
+    for (const session of this.sessions) {
+      this.deleteSession(session)
+    }
+  }
+
+  addTask (resolve, reject) {
+    let settled = false
+    const callback = (err, session) => {
+      if (!settled) {
+        settled = true
+        clearTimeout(timeoutId)
+        if (err) {
+          reject(err)
+        } else {
+          resolve(session)
+        }
+      }
+    }
+    const timeoutId = setTimeout(() => {
+      const index = this.queue.indexOf(callback)
+      if (index > -1) {
+        this.queue.splice(index, 1)
+      }
+      callback(new TimeoutError(`No session could be aquired within ${this.connectTimeout} ms`))
+    }, this.connectTimeout)
+    this.queue.push(callback)
+  }
+
+  acquire () {
+    const ticket = new Promise((resolve, reject) => this.addTask(resolve, reject))
+    this.dispatch()
+    return ticket
+  }
+
+  async dispatch () {
+    let scalingUp = false
+    const scaleUp = async () => {
+      if (!scalingUp) {
+        scalingUp = true
+        try {
+          await this.createSession()
+        } catch (err) {
+          const callback = this.queue.shift()
```

Comment:
if create session fails dispatch the error to all pending requests in the queue.

Reply:
✔️